File: Mellow/Mellow/Protocol/Commands.swift
import Foundation

// control plane (PROTOCOL.md §6). all multi-byte fields LE. written to cmd char (write-without-response)
enum RingCommand {

    // startup probe 08 03 00 00 00 (§6.1 step 1)
    static let timeOrIdReq: [UInt8] = [0x08, 0x03, 0x00, 0x00, 0x00]

    // sec_cfg pre / neg (§6.1 steps 3,5)
    static let secCfgPre: [UInt8] = [0x2f, 0x02, 0x01, 0x00]
    static let secCfgNeg: [UInt8] = [0x2f, 0x02, 0x01, 0x01]

    // request handshake nonce 2f 01 2b (§6.1 step 7)
    static let requestNonce: [UInt8] = [0x2f, 0x01, 0x2b]

    // submit handshake proof 2f 11 2d <proof:16> (§6.1 step 10)
    static func submitProof(_ proof: [UInt8]) -> [UInt8] {
        precondition(proof.count == 16)
        return [0x2f, 0x11, 0x2d] + proof
    }

    // subscribe enable/disable 16 01 0X (§6.6)
    static let subscribeEnable: [UInt8] = [0x16, 0x01, 0x02]
    static let subscribeDisable: [UInt8] = [0x16, 0x01, 0x00]

    // state_cmd 1c 01 bf, engage data plane (§6.1 step 13)
    static let stateCmd: [UInt8] = [0x1c, 0x01, 0xbf]

    // battery probe 0c 00 (§6.1 step 14)
    static let batteryReq: [UInt8] = [0x0c, 0x00]

    // capability dance reads/writes (§4.3, §6.1 step 15)
    static let capabilityDance: [[UInt8]] = [
        [0x2f, 0x02, 0x20, 0x02], [0x2f, 0x02, 0x20, 0x04], [0x2f, 0x02, 0x03, 0x01],
        [0x2f, 0x02, 0x20, 0x0b], [0x2f, 0x02, 0x20, 0x0d], [0x2f, 0x02, 0x20, 0x03],
        [0x2f, 0x02, 0x20, 0x0b], [0x2f, 0x02, 0x20, 0x10],
    ]

    // data_flush 28 01 00 (§6.5)
    static let dataFlush: [UInt8] = [0x28, 0x01, 0x00]

    // GetEvent 10 09 <ringTimestamp:4 LE> <max_events:1> <flags:4 LE> (§6.4)
    static func getEvent(cursor: UInt32, maxEvents: UInt8, flags: UInt32 = 0xFFFFFFFF) -> [UInt8] {
        var b: [UInt8] = [0x10, 0x09]
        b.append(UInt8(cursor & 0xff))
        b.append(UInt8((cursor >> 8) & 0xff))
        b.append(UInt8((cursor >> 16) & 0xff))
        b.append(UInt8((cursor >> 24) & 0xff))
        b.append(maxEvents)
        b.append(UInt8(flags & 0xff))
        b.append(UInt8((flags >> 8) & 0xff))
        b.append(UInt8((flags >> 16) & 0xff))
        b.append(UInt8((flags >> 24) & 0xff))
        return b
    }

    // time sync 12 09 <token> <counter:3 LE> 00 00 00 00 f6 (§6.3). counter = unix_s / 256
    static func timeSync(unixSeconds: Int, token: UInt8 = 0x01) -> [UInt8] {
        let counter = unixSeconds / 256
        return [0x12, 0x09, token,
                UInt8(counter & 0xff), UInt8((counter >> 8) & 0xff), UInt8((counter >> 16) & 0xff),
                0x00, 0x00, 0x00, 0x00, 0xf6]
    }

    // DHR burst trigger sequence (§6.7), force live high-rate HR sampling
    static let dhrBurst: [[UInt8]] = [
        [0x2f, 0x02, 0x20, 0x02],   // read DHR (capability check)
        [0x2f, 0x03, 0x22, 0x02, 0x03], // write DHR.byte_0 = 3 (burst)
        [0x2f, 0x03, 0x26, 0x02, 0x02], // write DHR.byte_2 = 2 (sub-mode)
    ]

    // MARK: - Live PPG / Daytime-HR streaming (§6.7, hr_live.py)

    // event subscribe 18 03 <cat> <flagLo> <flagHi>, sent once after handshake to open
    // the data categories live HR push needs. (cat, flag) pairs verbatim from proven Python flow
    static let liveFeatureCategories: [(cat: UInt8, flag: UInt16)] = [
        (0x14, 0x1000), (0x18, 0x1000), (0x28, 0x0900),
        (0x34, 0x0400), (0x04, 0x1000), (0x08, 0x1000),
    ]

    // 18 03 <cat> <flagLo> <flagHi> event-subscribe frame
    static func eventSubscribe(category: UInt8, flag: UInt16) -> [UInt8] {
        [0x18, 0x03, category, UInt8(flag & 0xff), UInt8(flag >> 8)]
    }

    // set feature mode 2f 03 22 <featureID> <mode>. DaytimeHR feature 0x02,
    // mode 0x03 = Requested-subscription. send ONCE to begin PPG push stream
    static func setFeatureMode(feature: UInt8, mode: UInt8) -> [UInt8] {
        [0x2f, 0x03, 0x22, feature, mode]
    }

    // set feature subscription 2f 03 26 <featureID> <subscription>. DaytimeHR
    // feature 0x02, subscription 0x02 = Latest. send ONCE, right after mode
    static func setFeatureSubscription(feature: UInt8, subscription: UInt8) -> [UInt8] {
        [0x2f, 0x03, 0x26, feature, subscription]
    }

    static let daytimeHRFeature: UInt8 = 0x02

    // two writes that begin live PPG push stream. set ONCE then leave ring
    // uninterrupted (re-sending repeatedly stalls the stream)
    static var activateDaytimeHR: [[UInt8]] {
        [setFeatureMode(feature: daytimeHRFeature, mode: 0x03),       // 2f 03 22 02 03
         setFeatureSubscription(feature: daytimeHRFeature, subscription: 0x02)] // 2f 03 26 02 02
    }

    // feature state read 2f 02 20 <featureID>, observe-only keep-alive poll while
    // streaming (does NOT re-trigger measurement)
    static func readFeatureState(feature: UInt8) -> [UInt8] {
        [0x2f, 0x02, 0x20, feature]
    }

    // MARK: - Workout / Exercise-HR streaming (RE'd from official strength-training capture)
    //
    // dense live-HR trigger during a logged workout is Exercise-HR feature (0x03)
    // in mode Requested (0x02), NOT DaytimeHR (0x02) and NOT 0x06. paired with two
    // Data-Collection enables and a repeated Get-Events poll. records come back as
    // normal inner records on notify char (0x80 green IBI, 0x60 IBI+amp, 0x46 temp,
    // 0x47 motion, 0x7e/0x7f steps), handled by existing decode path

    static let exerciseHRFeature: UInt8 = 0x03

    static let exerciseHRModeRequested: UInt8 = 0x02   // START
    static let exerciseHRModeOff: UInt8 = 0x00         // STOP

    // data collection 03 01 <arg> (opcode 0x03, len 0x01, arg byte).
    // arg 0x03 = Exercise HR, arg 0x0b = Real Steps
    static func dataCollection(_ arg: UInt8) -> [UInt8] { [0x03, 0x01, arg] }

    static let dataCollectionExerciseHR: UInt8 = 0x03
    static let dataCollectionRealSteps: UInt8 = 0x0b

    // workout get-events poll 10 09 <startTs:4 LE> ff ff ff ff ff (maxEvents 0xff,
    // flags 0xffffffff). official app repeats ~every 4 s to pull workout records
    static func getEvents(since: UInt32) -> [UInt8] {
        getEvent(cursor: since, maxEvents: 0xff, flags: 0xffffffff)
    }

    // START sequence for dense workout/Exercise-HR streaming, exact capture order:
    //   2f 03 22 03 02   set feature mode: Exercise-HR (0x03), Requested (0x02)
    //   03 01 03         data collection: Exercise HR
    //   03 01 0b         data collection: Real Steps
    //   16 01 02         subscribe enable
    //   18 03 18 00 10   event subscribe: category 0x18, flag 0x1000
    //   28 01 00         data flush
    static var activateWorkoutHR: [[UInt8]] {
        [setFeatureMode(feature: exerciseHRFeature, mode: exerciseHRModeRequested), // 2f 03 22 03 02
         dataCollection(dataCollectionExerciseHR),                                  // 03 01 03
         dataCollection(dataCollectionRealSteps),                                   // 03 01 0b
         subscribeEnable,                                                           // 16 01 02
         eventSubscribe(category: 0x18, flag: 0x1000),                              // 18 03 18 00 10
         dataFlush]                                                                 // 28 01 00
    }

    // STOP frame: set feature mode Exercise-HR (0x03) Off (0x00) = 2f 03 22 03 00
    static var stopWorkoutHR: [UInt8] {
        setFeatureMode(feature: exerciseHRFeature, mode: exerciseHRModeOff)
    }

    // MARK: - Provisioning (provision.py)

    // enrollment key-set 24 10 <auth_key:16>, writes our own auth_key into a
    // factory-reset ring (account-free provisioning). unused by live UI
    static func keySet(_ authKey: [UInt8]) -> [UInt8] {
        precondition(authKey.count == 16)
        return [0x24, 0x10] + authKey
    }

    // MARK: - Onboarding + Live HR (captured from official Oura app, 2026-06-16)
    //
    // right after key-set + handshake, official app runs a feature-enable sequence
    // that flips ring into "fully onboarded" (tracks HR/sleep), then triggers Live
    // HR with a single 0x06 write. live data streams back as 0x33 records (6 int16
    // PPG samples each) plus occasional 0x80 green-IBI records. verified on our key

    // post-key-set onboarding sequence official app sends (handle 0x15), in capture
    // order. enables full sensor feature set (0x02 DaytimeHR, 0x0b RealSteps,
    // 0x03 ExerciseHR, 0x04 SpO2, 0x0d CVA/rawPPG, 0x12)
    static let onboardingSequence: [[UInt8]] = [
        [0x2f, 0x02, 0x03, 0x01],         // sec/config step 0x03 (was missing)
        [0x1c, 0x01, 0xff],               // state_cmd (ff, not bf)
        [0x18, 0x03, 0x18, 0x00, 0x10],   // event_subscribe cat 0x18
        [0x08, 0x03, 0x00, 0x00, 0x00],   // time_or_id_req
        [0x2f, 0x03, 0x22, 0x02, 0x03],   // set feature 0x02 mode 0x03
        [0x2f, 0x03, 0x22, 0x0b, 0x01],   // set feature 0x0b mode 0x01
        [0x2f, 0x03, 0x26, 0x02, 0x02],   // feature 0x02 subscription Latest
        [0x2f, 0x03, 0x22, 0x03, 0x01],   // set feature 0x03 mode 0x01
        [0x16, 0x01, 0x02],               // subscribe enable
        [0x2f, 0x03, 0x22, 0x12, 0x01],   // set feature 0x12 mode 0x01
        [0x2f, 0x03, 0x22, 0x04, 0x01],   // set feature 0x04 (SpO2) mode 0x01
        [0x28, 0x01, 0x00],               // data_flush
        [0x2f, 0x03, 0x22, 0x0d, 0x01],   // set feature 0x0d (raw PPG) mode 0x01
    ]

    // measure HR trigger 06 04 20000000 (0x06 Set Realtime Measurements). fires
    // dense live PPG stream (0x33 records). re-send to keep/refresh measurement
    static let measureHeartRate: [UInt8] = [0x06, 0x04, 0x20, 0x00, 0x00, 0x00]

    // soft reset 0e 01 ff (§6.8), triggers ring reboot 22-35 s later
    static let softReset: [UInt8] = [0x0e, 0x01, 0xff]

    // factory reset 1a 00 (opcode 0x1A, len 0x00, no payload). ring replies
    // 1b <status:u16 LE> (0 = ok), wipes stored auth_key, reboots blank so official
    // app can reclaim it. see RingSession.factoryReset()
    static let factoryReset: [UInt8] = [0x1a, 0x00]
}

File: oura_core/src/commands.rs
//! Control-plane outer-frame builders.
//!
//! exact byte sequences the phone writes to the ring command characteristic.
//! pure functions (host supplies nonces/timestamps), bytes reproducible.
//! every sequence verified against captured phone<->ring traffic

use crate::crypto::{compute_handshake_proof, CryptoError};

// handshake (secure session, opcode 0x2f)

/// phone -> ring handshake start frame `2F 01 2B`
pub fn handshake_start() -> [u8; 3] {
    [0x2f, 0x01, 0x2b]
}

/// from the ring's 15-byte nonce (body of `2F 10 2C <nonce:15>`) and 16-byte
/// `auth_key`, build proof frame `2F 11 2D <proof:16>` (19 bytes)
pub fn handshake_proof_frame(
    auth_key: &[u8],
    nonce: &[u8],
) -> Result<Vec<u8>, CryptoError> {
    let proof = compute_handshake_proof(auth_key, nonce)?;
    let mut frame = Vec::with_capacity(3 + 16);
    frame.extend_from_slice(&[0x2f, 0x11, 0x2d]);
    frame.extend_from_slice(&proof);
    Ok(frame)
}

// time sync (opcode 0x12)

/// time-sync request frame:
///
/// ```text
/// 12 09 <token:1> <counter:3 LE> 00 00 00 00 f6
/// ```
///
/// counter = unix_time_s / 256. token is a host-supplied random byte (ring does
/// not validate it)
pub fn time_sync_frame(token: u8, unix_time_s: u64) -> [u8; 11] {
    let counter = (unix_time_s / 256) as u32;
    [
        0x12,
        0x09,
        token,
        (counter & 0xff) as u8,
        ((counter >> 8) & 0xff) as u8,
        ((counter >> 16) & 0xff) as u8,
        0x00,
        0x00,
        0x00,
        0x00,
        0xf6,
    ]
}

// parameter RPC (opcode 0x2f feature/param plane)

/// documented parameter IDs
pub mod param {
    /// Daytime Heart Rate, bytes 0/2 are mode/sub-mode
    pub const DHR: u8 = 0x02;
    /// activity HR enable, byte 0 toggle
    pub const ACTIVITY_HR: u8 = 0x03;
    /// SpO2 enable, byte 0 toggle
    pub const SPO2: u8 = 0x04;
    /// companion to ACTIVITY_HR, read-only in observed traffic
    pub const ACTIVITY_HR_AUX: u8 = 0x0B;
    pub const UNMAPPED_0D: u8 = 0x0D;
    pub const UNMAPPED_10: u8 = 0x10;
    /// polled by the official app at connection, answers all zeroes
    pub const UNMAPPED_12: u8 = 0x12;
}

/// the param sweep the official app performs once per connection, in order.
/// not required to start a measurement; recorded so the connection flow can be
/// matched against the app when diagnosing a ring that withholds features.
pub const CONNECT_PARAM_SWEEP: [u8; 5] = [0x12, 0x0c, 0x0b, 0x04, 0x10];

/// `2F 02 20 <param>` request the 4-byte param value
pub fn read_param(param_id: u8) -> [u8; 4] {
    [0x2f, 0x02, 0x20, param_id]
}

/// `2F 03 22 <param> <value>` set BYTE 0 of the param
pub fn write_param_byte0(param_id: u8, value: u8) -> [u8; 5] {
    [0x2f, 0x03, 0x22, param_id, value]
}

/// `2F 03 26 <param> <value>` set BYTE 2 of the param
pub fn write_param_byte2(param_id: u8, value: u8) -> [u8; 5] {
    [0x2f, 0x03, 0x26, param_id, value]
}

/// byte-perfect Daytime-HR mode write: byte-0 (mode) then byte-2 (sub-mode).
/// frames in send order.
///
/// no `read_param` prefix. a capture of the official app on a second ring
/// (2026-08-04, see EXTERNAL_CAPTURE_FINDINGS.md) shows it starts a measurement
/// with exactly these two writes. it does issue `read_param` sweeps, but at
/// connection time over params 0x12/0x0c/0x0b/0x04/0x10, unrelated to starting
/// a measurement.
pub fn set_dhr_mode(mode: u8, sub_mode: u8) -> Vec<Vec<u8>> {
    vec![
        write_param_byte0(param::DHR, mode).to_vec(),
        write_param_byte2(param::DHR, sub_mode).to_vec(),
    ]
}

/// on-demand HR burst: DHR mode=3 / sub-mode=2
pub fn request_hr_on_demand() -> Vec<Vec<u8>> {
    set_dhr_mode(3, 2)
}

/// stop an on-demand burst the way the app does: mode back to 1 (automatic),
/// subscription back to 0 (off). the ring also reverts on its own after ~20 s.
pub fn stop_hr_on_demand() -> Vec<Vec<u8>> {
    set_dhr_mode(1, 0)
}

/// toggle SpO2 sampling (byte-0 write to the SpO2 param)
pub fn set_spo2(on: bool) -> [u8; 5] {
    write_param_byte0(param::SPO2, if on { 0x01 } else { 0x00 })
}

pub fn set_activity_hr(on: bool) -> [u8; 5] {
    write_param_byte0(param::ACTIVITY_HR, if on { 0x01 } else { 0x00 })
}

// history fetch / reset / subscribe

/// phone -> ring `10 09 <ring_timestamp:4 LE> <max_events:1> <flags:4 LE>` 11
/// bytes. ring streams every event with `ringTimestamp > ring_timestamp`
pub fn request_events_since(ring_timestamp: u32, max_events: u8, flags: u32) -> [u8; 11] {
    [
        0x10,
        0x09,
        (ring_timestamp & 0xff) as u8,
        ((ring_timestamp >> 8) & 0xff) as u8,
        ((ring_timestamp >> 16) & 0xff) as u8,
        ((ring_timestamp >> 24) & 0xff) as u8,
        max_events,
        (flags & 0xff) as u8,
        ((flags >> 8) & 0xff) as u8,
        ((flags >> 16) & 0xff) as u8,
        ((flags >> 24) & 0xff) as u8,
    ]
}

/// default catch-up: from `ring_timestamp`, up to 255 events, all flags
pub fn request_events_since_default(ring_timestamp: u32) -> [u8; 11] {
    request_events_since(ring_timestamp, 255, 0xFFFF_FFFF)
}

/// soft reset: phone sends `0E 01 FF`, ring acks `0F 01 00` and reboots
pub fn soft_reset() -> [u8; 3] {
    [0x0e, 0x01, 0xff]
}

/// per-category event-subscribe frame `18 03 <category:u8> <flags:u16 LE>`
pub fn event_subscribe(category: u8, flags: u16) -> [u8; 5] {
    [
        0x18,
        0x03,
        category,
        (flags & 0xff) as u8,
        ((flags >> 8) & 0xff) as u8,
    ]
}

/// subscribe-enable toggle `16 01 02`
pub fn subscribe_enable() -> [u8; 3] {
    [0x16, 0x01, 0x02]
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn handshake_start_bytes() {
        assert_eq!(handshake_start(), [0x2f, 0x01, 0x2b]);
    }

    #[test]
    fn proof_frame_shape() {
        let key = [0u8; 16];
        let nonce = [0u8; 15];
        let f = handshake_proof_frame(&key, &nonce).unwrap();
        assert_eq!(f.len(), 19);
        assert_eq!(&f[..3], &[0x2f, 0x11, 0x2d]);
    }

    #[test]
    fn time_sync_trailer_is_f6() {
        let f = time_sync_frame(0xAB, 256 * 5);
        assert_eq!(f[0], 0x12);
        assert_eq!(f[1], 0x09);
        assert_eq!(f[2], 0xAB);
        assert_eq!(&f[3..6], &[5, 0, 0]);
        assert_eq!(f[10], 0xf6);
    }

    #[test]
    fn get_event_frame() {
        let f = request_events_since_default(0x0083_dd11);
        assert_eq!(f[0], 0x10);
        assert_eq!(f[1], 0x09);
        assert_eq!(&f[2..6], &[0x11, 0xdd, 0x83, 0x00]);
        assert_eq!(f[6], 255);
        assert_eq!(&f[7..11], &[0xff, 0xff, 0xff, 0xff]);
    }

    #[test]
    fn dhr_burst_sequence() {
        // exactly what the official app writes to start a measurement
        let seq = request_hr_on_demand();
        assert_eq!(seq.len(), 2);
        assert_eq!(seq[0], vec![0x2f, 0x03, 0x22, 0x02, 0x03]);
        assert_eq!(seq[1], vec![0x2f, 0x03, 0x26, 0x02, 0x02]);
    }

    #[test]
    fn dhr_stop_sequence() {
        let seq = stop_hr_on_demand();
        assert_eq!(seq.len(), 2);
        assert_eq!(seq[0], vec![0x2f, 0x03, 0x22, 0x02, 0x01]);
        assert_eq!(seq[1], vec![0x2f, 0x03, 0x26, 0x02, 0x00]);
    }
}
